src/pop/blockchain/alt_block_tree.cpp

```diff
@@ -468,10 +468,28 @@ AltBlockTree::BlockPayloadMutator::BlockPayloadMutator(tree_t& tree,
   appendIds<VbkBlock>(ids_, block_);
 }
 
-bool AltBlockTree::BlockPayloadMutator::isStatefulDuplicate(
-    const id_vector_t& payload_id) {
-  // make sure existing blocks do not contain this id
-  for (const auto& hash : payload_index_.getContainingAltBlocks(payload_id)) {
+template <>
+bool AltBlockTree::BlockPayloadMutator::isStatefulDuplicate<ATV>(
+    const ATV::id_t& atv_id) {
+  auto id = atv_id.asVector();
+  // make sure existing blocks from the active chain do not contain this id
+  for (const auto& hash : payload_index_.getContainingAltBlocks(id)) {
+    const auto* candidate = tree_.getBlockIndex(hash);
+    if (chain_.contains(candidate)) {
+      // duplicate found in 'candidate'
+      return true;
+    }
+  }
+
+  return false;
+}
+
+template <>
+bool AltBlockTree::BlockPayloadMutator::isStatefulDuplicate<VTB>(
+    const VTB::id_t& vtb_id) {
+  auto id = vtb_id.asVector();
+  // make sure existing blocks from the active chain do not contain this id
+  for (const auto& hash : payload_index_.getContainingAltBlocks(id)) {
     const auto* candidate = tree_.getBlockIndex(hash);
     if (chain_.contains(candidate)) {
```

Comment:
For VTB: find "containing block" and search for ID. If id exists -> it's duplicate.

Reply:
and we need to add new validation - check if these blocks (VTB containing, block of proof, ATV block of proof) are behind final - if they are, then they should fail validation